src/native/eventpipe/ep-buffer-manager.c

```diff
@@ -1231,6 +1256,61 @@ ep_buffer_manager_write_all_buffers_to_file_v4 (
+		EP_SPIN_LOCK_ENTER (&buffer_manager->rt_lock, section4)
+			if (buffer_manager_try_peek_sequence_point (buffer_manager, &sequence_point)) {
+				// foreach (sequence_point in buffer_manager->sequence_point_list)
+				for (ep_rt_sequence_point_list_iterator_t sequence_point_list_iterator = ep_rt_sequence_point_list_iterator_begin(&buffer_manager->sequence_points);
+					!ep_rt_sequence_point_list_iterator_end(&buffer_manager->sequence_points, &sequence_point_list_iterator);
+					ep_rt_sequence_point_list_iterator_next(&sequence_point_list_iterator)) {
+
+					sequence_point = ep_rt_sequence_point_list_iterator_value(&sequence_point_list_iterator);
+
+					// foreach (session_state in session_states_to_delete)
+					for (ep_rt_thread_session_state_array_iterator_t thread_session_state_array_iterator = ep_rt_thread_session_state_array_iterator_begin(&session_states_to_delete);
+						!ep_rt_thread_session_state_array_iterator_end(&session_states_to_delete, &thread_session_state_array_iterator);
+						ep_rt_thread_session_state_array_iterator_next(&thread_session_state_array_iterator)) {
+
+						EventPipeThreadSessionState * thread_session_state = ep_rt_thread_session_state_array_iterator_value(&thread_session_state_array_iterator);
+						uint32_t thread_sequence_number = 0;
+						bool exists = ep_rt_thread_sequence_number_map_lookup (ep_sequence_point_get_thread_sequence_numbers_cref (sequence_point), thread_session_state, &thread_sequence_number);
+						if (exists) {
+							ep_rt_thread_sequence_number_map_remove (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), thread_session_state);
+							// every entry of this map was holding an extra ref to the thread (see: ep-event-instance.{h|c})
+							ep_thread_release (ep_thread_session_state_get_thread (thread_session_state));
+						}
+					}
+				}
+			}
+
+		EP_SPIN_LOCK_EXIT (&buffer_manager->rt_lock, section4)
+	}
+
+	// foreach (session_state in session_states_to_delete)
+	for (ep_rt_thread_session_state_array_iterator_t thread_session_state_array_iterator = ep_rt_thread_session_state_array_iterator_begin(&session_states_to_delete);
+			!ep_rt_thread_session_state_array_iterator_end(&session_states_to_delete, &thread_session_state_array_iterator);
+			ep_rt_thread_session_state_array_iterator_next(&thread_session_state_array_iterator)) {
+
+		EventPipeThreadSessionState * thread_session_state = ep_rt_thread_session_state_array_iterator_value(&thread_session_state_array_iterator);
+		EP_ASSERT (thread_session_state != NULL);
+		// This may be the last reference to a given EventPipeThread, so make a ref to keep it around till we're done
+		EventPipeThreadHolder thread_holder;
+		if (ep_thread_holder_init (&thread_holder, ep_thread_session_state_get_thread (thread_session_state))) {
+
+			// briefly hold the thread lock to check it's state
+			ep_rt_spin_lock_handle_t *thread_lock = ep_thread_get_rt_lock_ref (ep_thread_session_state_get_thread (thread_session_state));
```

Comment:
More clean if you get thread instance from the thread holder instead of mixing thread holder and thread session state when accessing EventPipeThread in this code section. Might be even more cleaner (and less code) to put EventPiptThread into a local variable and use that.

Reply:
a remnant of refactoring the for loops several times.

Reply:
The holder is needed cause this may be the last reference to the EventPipeThread so we need to have a strong reference to keep it alive till we finish.